### ISSUE TITLE
Is it also an idea to write a real ansible module for MQ?

### ISSUE THREAD

**Issue by ouwe-knutselaar** [ibm-messaging/mq-ansible-yaml-commands#1]
I see in the code that you use the API. However, this is not always available due to the fact that you have to open an extra port on the MQ system.
So is it an idea to also add an Ansible module to parse the commands?

**Comment by mayur-raja** [comment]
Hi Ouwe,

Thank you for your question and apologies for the delay in my response. I had overlooked the Issues tab and only just noticed it. 

1. We did consider writing an MQ module but the general concensus in the MQ team was that this would require long term effort to keep up to date with changes in base MQ. So, for now, we have decided aganst creating an MQ module. Though, I think we may need to reconsider this at a later date as more customers start to use Ansible for automation and raise new requirements. 

2. We opted for the MQ Admin REST API as this would allow us to provide a solution that works on distributed as well as z/OS platforms. However, I see the concerns you raise. IBM does provide an ibm_zos_core collection which contains a zos_operator module. I have used this in playbooks to direct MQSC commands to a IBM MQ for z/OS Queue Manager. Connectivity to the z/OS LPAR can be established via SSH, and there is no dependency on the mqweb server. Let me know if this would help you and I will discuss with the MQ architects to see if they agree with us providing sample playbooks for this.

Regards .. Mayur

**Comment by mamwl1** [comment]
There are already samples for the z/OS scenario mentioned in point 2: https://community.ibm.com/community/user/integration/blogs/matthew-leming1/2021/05/21/using-ansible-to-orchestrate-configuration-of-ibm

**Comment by mayur-raja** [comment]
Yes Matt, good point! Thank you, and apologies but I overlooked those samples because I was currently focusing on the YAML ones.

**Comment by ouwe-knutselaar** [comment]
Via SSH is for us the best option.
We don’t have z/OS anymore but we still have Linux and HPNSK.

I also was building MQ modules in python with limited functionality last year. It are in the background always a wrapper around runmqsc but it worked very well.

**Comment by mayur-raja** [comment]
For distributed MQ platforms, we did write playbooks that invoke the ansible shell module to pipe commands into runmqsc, like this: 

 `- name: Execute a MQ Command`
        `  shell:`
          `    cmd:  "echo 'DISPLAY QMGR ALL' | ./runmqsc MQ01"`
          `    chdir: /opt/mqm/bin`
        `  register: mqdispqmgr`
   
This allows us to issue commands without having to code a dedicated MQ module in Python. However, I do agree with you that to have a dedicated MQ module would provide more flexibility, allow complexity to be hidden within the module, and allow for a declarative approach for provisioning of resources with the use of schemas. Actually, we did write a simple module and get it running but we opted for the REST API approach as it allowed us to more readily run commands against both distributed and z/OS platforms.
